### PR TITLE
haskell-ledger-bindings, test connect against sandbox

### PR DESCRIPTION
**Pull request by nickchapman-da**
Main focus of this PR is to provide the framework code (Sandbox.hs) for spinning up a sandbox ledger on a dynamically selected port, which the haskell testsuite can make use of.

Tests for connect & ledger-identity service. (Tests.hs)

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/hs/bindings/BUILD.bazel [hunk 1]
```diff
@@ -36,3 +37,36 @@ da_haskell_binary(
         "//language-support/hs/bindings:hs-ledger",
     ],
 )
+
+daml_compile(
+    name = "quickstart",
+    srcs = glob(["test/daml/quickstart/*.daml"]),
+    main_src = "test/daml/quickstart/Main.daml",
+    target = "1.1",
```
Comment (neil-da) [inline language-support/hs/bindings/BUILD.bazel]: Why not 1.3?

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 2]
```diff
@@ -0,0 +1,123 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Sandbox ( -- Run a sandbox for testing on a dynamically selected port
```
Comment (neil-da) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Why not `DA.Ledger.Sandbox` as the module name? Having the module name and path not mix is often a poor idea.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    SandboxSpec(..),
+    Sandbox(port),
+    startSandbox,
+    shutdownSandbox,
+    withSandbox
+    ) where
+
+import           Control.Concurrent.Async (race)
+import           Control.Monad            (when)
```
Comment (neil-da) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: No need to give an export list for this and many of the things below.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+withSandbox :: SandboxSpec -> (Sandbox -> IO a) -> IO a
+withSandbox spec f =
+    bracket (startSandbox spec)
+    shutdownSandbox
+    f
+
+data Timeout = Timeout deriving Show
+instance Exception Timeout
+
+timeout :: Seconds -> String -> IO a -> IO a
```
Comment (neil-da) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Why not just use `System.Time.Extra.timeout`? Inventing your own timeout handler is not a good plan.

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (cocreature) [reply language-support/hs/bindings/BUILD.bazel]: You could also just not specify it in which case the default of the compiler will be used which has the advantage that we notice breakages when we upgrade that.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+module Sandbox ( -- Run a sandbox for testing on a dynamically selected port
+    SandboxSpec(..),
+    Sandbox(port),
+    startSandbox,
+    shutdownSandbox,
+    withSandbox
+    ) where
+
+import           Control.Concurrent.Async (race)
```
Comment (cocreature) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: We stopped aligning imports to minimize diffs. But we haven’t gone through the codebase and changed all old imports so you might see some inconsistencies here.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+selectedPort :: Int
+selectedPort = 0 --dynamic port selection
+
+sandboxProcess :: SandboxSpec -> CreateProcess
+sandboxProcess SandboxSpec{dar} =
+    proc binary [ dar, "--port", show selectedPort]
+    where
+        binary = "ledger/sandbox/sandbox-binary"
+
+startSandboxProcess :: SandboxSpec -> IO (ProcessHandle,Maybe Handle)
```
Comment (cocreature) [inline language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: It is probably a good idea to use `withCreateProcess` here.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 5 above)
Comment (neil-da) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: In truth we just fixed it in GHC upstream, so are desperately waiting for qualified after the module name! Until then, anything goes.

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/BUILD.bazel]: I removed the version

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 2 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Oversight. Fixed.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 4 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Changed to use `System.Time.Extra.timeout` as suggested.
But this only replaces the use of `race`. The rest of the wrapper (now named `timeoutError`) persists.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 5 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: I unaligned my imports. No more `stylish-haskell` for me!

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 6 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: Having separate `startSandbox` and `shutdownSandbox`, we can't use `withCreateProcess`.
But I probably should make use of the standard `cleanupProcess`.

File: language-support/hs/bindings/test/DA/Ledger/Sandbox.hs (hunk 3 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/test/DA/Ledger/Sandbox.hs]: I read a suggestion on Hackage that imports should take either the form: `import qualified M` or `import M(...)`, so it's clear where the definition for every identifier is coming from.

As a Haskell Noob, I like very much being this explicit, especially while I am still finding my way around the standard libraries.

Of course, I accept that some standard definitions (`when` ?) are so pervasive that perhaps the rule doesn't need to apply to them.

**Pull request merged**